### PR TITLE
perf: hydrate SSR subtrees in a single pre-order pass

### LINKED ISSUE #431: [Bug]: An `<if>` inside an `<if>` is anchored in the parent's coordinate space, so its update lands in an unrelated element
### Before filing

- [x] I searched existing issues and did not find a duplicate. The closest is #405 / #406 ("A sibling `<for>` after a root-level nested repeat hydrates against the wrong marker"), which is the same class of defect for repeats. This is the `<if>` case, and it still reproduces on 0.0.22.
- [x] This is not a security vulnerability.

### Area

WebUI Framework package

### Summary

A conditional nested inside another conditional compiles to an anchor path expressed in the **component's** coordinate space rather than its own branch's, so it can collide with an anchor belonging to an unrelated subtree. When the inner condition changes on the client, the content is inserted into whatever element occupies that path - in the reproduction below a `<span>` that belongs in `.cell` is inserted into `<header class="head">` instead.

SSR is correct, and no hydration-mismatch warning is logged. Only the post-hydration update is misplaced.

### Minimal reproduction

`src/demo-panel/demo-panel.html` - all three parts matter; deleting the static `<section class="box">` (part 2) makes it render correctly:

```html
<section class="root">
  <!-- 1. a root-level conditional -->
  <if condition="showStats"><header class="head">head</header></if>

  <!-- 2. a STATIC element that itself contains conditionals -->
  <section class="box">
    <div class="box-head">
      <if condition="compact"><h1>compact title</h1></if>
      <if condition="!compact"><h2>full title</h2></if>
    </div>
    <div class="box-body">body</div>
    <if condition="compact"><div class="extra">extra</div></if>
  </section>

  <!-- 3. a root-level conditional holding a client-updating conditional -->
  <if condition="showStats">
    <section class="stats">
      <div class="grid">
        <div class="cell">
          <if condition="value"><span class="num">{{value}}</span></if>
          <if condition="!value"><span class="num pending">pending</span></if>
        </div>
      </div>
    </section>
  </if>
</section>
```

`src/demo-panel/demo-panel.ts` - the value arrives after hydration, as a measurement or a fetch would:

```ts
import { WebUIElement, observable } from '@microsoft/webui-framework';

export class DemoPanel extends WebUIElement {
  @observable value = '';

  override connectedCallback(): void {
    super.connectedCallback();
    setTimeout(() => { this.value = '42'; }, 50);
  }
}

DemoPanel.define('demo-panel');
```

`data/state.json`:

```json
{ "showStats": true, "compact": false }
```

`src/index.html` is a plain entry containing `<demo-panel></demo-panel>` and `<script type="module" src="/index.js">`; `src/index.ts` is `import './demo-panel/demo-panel.js';`.

### Steps to reproduce

1. `npm i -D @microsoft/webui@0.0.22 @microsoft/webui-framework@0.0.22 esbuild`
2. `npx esbuild src/index.ts --bundle --format=esm --target=es2022 --outfile=public/index.js` (with `experimentalDecorators: true` and `useDefineForClassFields: false`)
3. `npx webui serve ./src --state ./data/state.json --plugin=webui --port 8080 --servedir ./public`
4. Open `http://127.0.0.1:8080/`, wait ~1s for the observable to be set, then check where `.num` ended up:

```js
document.querySelector('demo-panel').shadowRoot.querySelector('.num').parentElement.className;
```

### Expected behavior

`.num` is inserted into `.cell`, replacing the `pending` placeholder - which is what happens when part 2 is deleted, or when the outer `<if>` around the `.stats` block is removed.

### Actual behavior

`.num` is inserted into `header.head`, and `.cell` is left empty:

```
{"numParent":"head","rootChildren":["header.head","section.box","section.stats"]}
```

The compiled template in the page's `webui-data` payload shows why - the inner conditional of branch 4 is emitted with the *same* anchor as a conditional belonging to the `.box` subtree:

```
root conditionals   (branch -> anchor, condition)
   0 -> [[0], 1]           showStats
   1 -> [[0, 1, 1], 1]     compact      <-- inside .box > .box-head
   2 -> [[0, 1, 1], 1, 1]  compact
   3 -> [[0, 1], 5]        compact
   4 -> [[0], 3]           showStats

branch 4 inner conditionals
   5 -> [[0, 1, 1], 1]     value        <-- identical anchor to root conditional 1
   6 -> [[0, 1, 1], 1, 1]  value
```

`[[0,1,1],1]` resolves inside the component root rather than inside branch 4's own content, so the update is applied to the first subtree that matches that path.

In a larger component with a few dozen conditionals the same defect also shows up as whole `<if>` branches rendering out of order relative to static siblings, so the two symptoms are worth checking together.

### Environment

- OS: macOS 26.6 (arm64)
- Rust version: n/a (reproduced through the `webui` CLI dev server)
- Node.js version: v24.18.0
- WebUI package or CLI version: 0.0.22 (also reproduces on 0.0.20)
- Command: `npx webui serve ./src --state ./data/state.json --plugin=webui --port 8080 --servedir ./public`

### Logs or terminal output

```shell
# after hydration on 0.0.22
{"numParent":"head","rootChildren":["header.head","section.box","section.stats"]}

# the SSR markup is correct - every placeholder is in the right place
<section class="root"><!--wc--><header class="head">head</header><!--/wc-->
<section class="box"><div class="box-head"><!--wc--><!--/wc--><!--wc--><h2>full title</h2><!--/wc--></div>
<div class="box-body">body</div><!--wc--><!--/wc--></section>
<!--wc--><section class="stats"><div class="grid"><div class="cell"><!--wc--><!--/wc--><!--wc--><!--/wc--></div></div></section><!--/wc--></section>
```

Nothing is logged when this happens - no hydration-mismatch warning - which is what makes it expensive to track down in an application: the page simply renders a value in the wrong element after hydration.

Workaround, in case it helps anyone hitting this: keep any conditional whose condition changes on the client as a direct child of the component root, and gate server-only regions with CSS (a bound `?data-*` attribute plus an attribute selector) rather than nesting them.

### PR DESCRIPTION
**Pull request by mohamedmansour**
Hydration adopted server-rendered DOM by resolving every binding independently: each one walked down from the component root, rescanning a parent's children on the way to find the Nth element. That makes hydration cost proportional to bindings times sibling count, so it degrades as a component gets wider.

A 50-node flat component with 50 bindings spent **2,269 sibling steps to bind 123 nodes** — 18.4 steps per node.

The compiler emits metadata in source order and the server renders in source order. One pre-order walk can therefore pair every template element with the element it hydrates, and hand out `<!--wc-->` / `<!--wr-->` markers in exactly the order the compiled `c` / `r` tables expect. Bindings resolve by lookup afterwards.

## Measured

Traversal steps are deterministic, so they are the primary metric — no timer noise.

| Fixture | Before | After | |
|---|---|---|---|
| `bench` (50 flat siblings, 65 bindings) | 2,269 | 128 | **17.7x fewer** |
| `hydration-bench` (300 instances) | 187,650 | 23,100 | **8.1x fewer** |

`bench` drops from 18.4 steps/node to 1.04 — quadratic to linear.

Wall clock, 150 instances, median of 5 runs:

| | Before | After |
|---|---|---|
| wide component | 8.9 ms | 6.6 ms |
| deep component | 2.7 ms | 2.2 ms |

Small components do a few more steps than before, because the walk is eager where the old lookup was lazy. Those are tens of steps on components that were already cheap, and the cost is now linear and predictable everywhere instead of quadratic in the worst case.

## Three hydration bugs, removed by construction

The same walk fixes a family of anchoring bugs, because a block's anchor is no longer reconstructed from a parent path plus a stateful scan cursor — it is simply the i-th marker.

1. **`Closes #431`** — a conditional nested inside another conditional was anchored in the parent's coordinate space, so its update landed in an unrelated element. A single scan cursor reset whenever the parent changed, so a block that returned to an earlier parent re-claimed a marker that parent had already used.
2. A conditional resumed scanning at its own `<!--wc-->` start, so the next sibling walked back into the branch just consumed and claimed a marker belonging to a block nested at that branch's root. The sibling then owned an unrelated range, and toggling it did nothing. Repeats already advanced past `<!--/wr-->`; conditionals did not.
3. Marker scanning walked straight through structural ranges, so a root-level `<for>` could claim the `<!--wr-->` of a `<for>` nested inside a preceding `<if>` branch. That **duplicated rows**, leaving the original server-rendered items orphaned.

Each is covered by a fixture that fails without its fix, verified by reverting each fix independently.

## Why hydration had no benchmark

The existing `bench` fixture measures `$update()` once a component is already live. Nothing measured the cost of adopting SSR DOM in the first place, which is why this went unnoticed. This PR adds a `hydration-bench` fixture that brackets `connectedCallback` across 150 instances each of a wide and a deep component and reports per-instance cost.

## Two constraints the implementation has to respect

Both were established by dumping real SSR output against compiled `h`, and both are now unit tests:

- The renderer **strips inter-element whitespace** that compiled `h` keeps. Element order aligns between template and SSR; text order does not. Only elements are paired.
- `<ul><for …></ul>` compiles to an **empty** `<ul>`, so "descend only where the template has children" misses markers. An element with no template children is still walked when the section has blocks to place. Child components are the exception — they contribute no children to the parent's `h`, so whatever the server rendered inside belongs to them.

## Maintenance

Net **-27 lines** in `template-element.ts`. `$findMarker`, both per-parent cursor maps, and the end-marker cursor are all deleted along with the class of bug they enabled.

## Element addressing

Locators used to carry a path of childNode offsets, so every binding descended that chain at runtime and both client paths needed ordinal reconciliation to line template offsets up with server output.

Elements are now numbered in **pre-order per compiled section** - `0` is the section root, elements are `1..N` in the order a depth-first walk of `h` meets them - and bindings are addressed by that number. Both runtimes rebuild the same numbering in one walk, so a binding resolves by array index.

```
before   "c":[[[0,["compact"]],1,[[0,1,1],1]]]
after    "c":[[[0,["compact"]],1,[5,1]]]
```

This is a **wire format break with no compatibility shim** - compiler and runtime ship together.

| | |
|---|---|
| `webui-data` across the fixture suite | 42,244 -> 41,264 bytes (**-2.3%**) |
| `template-element.ts` | **-46 lines** net |
| Deleted outright | `$resolve`, `$resolveSSR`, `emit_js_node_path` |

It also retires the subtlest part of the old contract: element ordinals had to be reconciled against SSR because the renderer strips inter-element whitespace that `h` keeps. Numbering removes the question. Text runs still resolve by ordinal, since text nodes genuinely do not line up.

Two constructs make the browser's parse of `h` disagree with the compiler's own tree, and both are accounted for: a bare `<tr>` gains an implied `<tbody>`, and `<template>` content is parked in `.content` where neither runtime walk reaches it. Everything else the parser does — `<p>` / `<li>` auto-closing and friends — changes nesting rather than sequence, and a pre-order index only cares about sequence.

Hydration wall clock is unchanged here, and that is the expected result - the asymptotic work was already removed above, and swapping a map lookup for an array read sits below this benchmark's noise floor. The gain is payload and a smaller contract, not speed.

## Cost characteristics

Hydrating one section is linear in the subtree it owns. Structural *nesting* is the exception and is not linear: each `<if>` / `<for>` hydrates a body the enclosing walk has already stepped over, so a chain of N nested blocks costs O(N²) marker steps.

| nesting depth | steps/instance |
|---|---|
| 4 | 26 |
| 8 | 82 |
| 16 | 290 |
| 32 | 1090 |

N is nesting depth, not node count, so a realistic depth of 8 costs ~82 steps for the whole chain. `test-hydration-nested` tracks this rather than paying for a cross-section marker-pairing cache that would add persistent state and plumbing to every nested hydrate call for a shape real templates do not reach.

## Verification

- 222 unit tests (+11 new, covering `skipBlockRange` and `buildSSRIndex` including the whitespace, empty-container and component-boundary cases)
- 169 e2e tests (+9 new), each structural fix verified by reverting it independently
- 536 Rust parser tests, including the ~30 assertions that encode the wire format
- `cargo xtask check` clean on all Rust stages
- `DESIGN.md` updated: the protocol tables, the new element-addressing rule, and the runtime contract

Closes #431

### CODE REVIEW

File: crates/webui-parser/src/plugin/webui.rs [hunk 1]
```diff
@@ -1750,11 +1744,11 @@ fn process_fragment_children(
                 }
             }
             FragmentNode::Element(element) => {
-                let mut element_path = parent_path.to_vec();
-                element_path.push(child_index);
+                *element_cursor += 1;
+                let element_index = *element_cursor;
```
[the remark this answers is not in the document]
Comment (mohamedmansour) [reply crates/webui-parser/src/plugin/webui.rs:1749]: Confirmed and fixed in 1a713511.

I checked both constructs against a real browser parse and both diverge as you describe:

```
<table><tr><td></td></tr></table><button>   →  TABLE,TBODY,TR,TD,BUTTON
<div><template><span/></template></div><button>  →  DIV,TEMPLATE,BUTTON
```

The compiler was emitting `button=4` in both cases against a real `5` and `3`.

One thing worth recording from checking the wider blast radius: the other recovery rules are harmless here. A pre-order index depends only on element *order*, not nesting, so `<p>`/`<li>` auto-closing and `<option>` reparenting all still agree. Only constructs that **add** an element (implied `tbody`) or **move one out of the light tree** (`template.content`) shift the sequence. That narrows the fix to exactly the two you named.

- Implied `<tbody>` is now emitted into `h` itself, so the browser has nothing to insert and both parses agree. Authored `<tbody>`/`<thead>` are not double-wrapped.
- `<template>` interiors are no longer numbered, matching both runtime walks. Bindings inside stay unreachable, exactly as they were before indices replaced node paths.

Covered by `test-parser-quirks` (e2e, real browser: table binding + parked template + a bound event after both) and four parser unit tests. I verified the e2e fails independently with each half of the fix reverted.

File: packages/webui-framework/src/element/markers.ts [hunk 2]
```diff
@@ -73,6 +73,176 @@ export function nextElement(marker: Comment): Element | null {
...
+    }
+
+    // The counter follows the template, never the server output, so a run of
+    // missing SSR elements leaves holes rather than shifting every index after
+    // it onto the wrong node.
+    index++;
+    if (s) {
+      elements[index] = s;
+      frame.s = s.nextSibling;
+    }
```
[the remark this answers is not in the document]
Comment (mohamedmansour) [reply packages/webui-framework/src/element/markers.ts:241]: Correct, and fixed in 1a713511.

I reproduced it with exactly the shape you described. The order of operations inside the loop is the cause: the SSR cursor advances *before* the template-exhausted check, so once the single-root template ran out, the scan walked past `<!--/wc-->` and collected the following sibling's `<!--wc-->`. That made `conds.length` 2 against `meta.c.length` 1, the count guard rejected the pairing, and the inner block fell back to slot positions - leaving the real inner range unwired.

Fixed by bounding the frame rather than reordering the checks. Reordering looked tempting but is wrong: a trailing `<if>` after the last element of a parent is legitimate and must still be collected, so popping as soon as the template is exhausted would lose it. Instead the entry frame carries a `solo` flag for in-place hydration, which is entered on exactly one element. That is safe precisely because the single-root path is guarded by `$hasRootStructuralSlot`, so a block with a section-root-level structural slot never takes it - there is no trailing marker at that level to miss.

Regression test added as requested: `test-conditional-block-escape` is `<if><section><if/></section></if><if/>`, and it fails on the previous commit.

File: packages/webui-framework/src/element/markers.ts [hunk 3]
```diff
@@ -73,6 +73,176 @@ export function nextElement(marker: Comment): Element | null {
...
+  const endTag = data === MARKER_COND_START ? MARKER_COND_END : MARKER_REPEAT_END;
+  let depth = 1;
+  let node: ChildNode | null = start.nextSibling;
+  while (node && depth > 0) {
+    if (node.nodeType === 8 /* COMMENT_NODE */) {
+      const d = (node as Comment).data;
+      if (d === data) depth++;
+      else if (d === endTag) depth--;
+    }
+    if (depth > 0) node = node.nextSibling;
```
[the remark this answers is not in the document]
Comment (mohamedmansour) [reply packages/webui-framework/src/element/markers.ts:94]: You are right on the complexity, and the linear-cost claim was overstated. I measured it with a chain of nested `<if>` blocks, counting traversal steps:

| depth | steps/instance | vs previous |
|---|---|---|
| 4 | 26 | - |
| 8 | 82 | 3.15x |
| 16 | 290 | 3.54x |
| 32 | 1090 | 3.76x |

Doubling depth multiplies steps by ~3.5-3.8x. Linear would be 2.0x, quadratic 4.0x - so it is quadratic, as you said.

I have added the structurally deep benchmark you asked for (`test-hydration-nested`, a chain of `<if>` blocks rather than nested static elements) and corrected the claim in both the code comment and the PR description.

I have **not** added the start/end pairing cache, and I want to be explicit that this is a judgement call rather than a disagreement. The N in O(N²) is nesting *depth*, not node count: at a realistic depth of 8 the entire chain costs ~82 steps, against ~30 if it were perfectly linear. Buying that back needs a marker-pairing map built once per component and threaded through every nested `$hydrate` call - persistent state and new plumbing on the hot path, to save tens of steps at depths real templates do not reach. Given the explicit goal of keeping this runtime small, tracking it with a benchmark seemed the better trade than paying for it unconditionally.

If the benchmark ever shows this mattering on a real app shape, the cache is the right fix and the measurement will be there to justify it.

File: packages/webui-framework/README.md [hunk 4]
```diff
@@ -785,36 +785,36 @@ stylesheet specifier for a component.
...
+### SSR resolution (`buildSSRIndex`)
 
-SSR DOM may differ from the compiled template — the browser's HTML parser can
-strip whitespace-only text nodes.  `$resolveSSR` walks the SSR DOM and the
-compiled template DOM **in parallel**, translating each child-node index into
-an element-ordinal or text-ordinal lookup:
+SSR DOM differs from the compiled template: the renderer strips inter-element
+whitespace, and `<if>` / `<for>` bodies are rendered inline between markers.
+`buildSSRIndex` walks the SSR DOM and the compiled template DOM **in lockstep**,
+skipping whole marker ranges, and numbers the result the same way:
```
[the remark this answers is not in the document]
Comment (mohamedmansour) [reply packages/webui-framework/README.md:810]: Correct - that paragraph contradicted the algorithm directly above it. Fixed in 1a713511.

It now distinguishes the two things it was conflating:

> This eliminates all *per-binding* annotation: no `data-w-*` attributes, no comment per text run, no DOM markers around individual bindings. What the SSR server does emit is the five structural comments documented above (`<!--wr-->`, `<!--wi-->`, `<!--/wr-->`, `<!--wc-->`, `<!--/wc-->`), which delimit `<if>` / `<for>` bodies and are what the walk skips over and anchors blocks on. They are removed once hydration completes.

I also swept the surrounding "path indices" wording in RENDERING.md, which described the addressing scheme this PR replaced.

**Review by Qusic — APPROVED** [review]
